File: ts/datastructure/otterop/datastructure/TestList.test.ts
import { Array } from '@otterop/lang/Array';
import { Generic } from '@otterop/lang/Generic';
import { String } from '@otterop/lang/String';
import { TestBase } from '@otterop/test/TestBase';
import { List } from './List';

export class TestList extends TestBase {

    public add() : void {
        let l : List<String> = new List<String>();
        l.add(String.wrap("a"));
        l.add(String.wrap("b"));
        l.add(String.wrap("c"));
        l.add(String.wrap("d"));
        l.add(String.wrap("e"));
        this.assertTrue(l.size() == 5, String.wrap("Size should be 5"));
    }

    public addRange() : void {
        let genericString : Generic<String> = new Generic<String>();
        let genericT : String = genericString.zero();
        let l : List<String> = new List<String>();
        let toAdd : Array<String> = Array.newArray(5, genericT);
        toAdd.set(0, String.wrap("a"));
        toAdd.set(1, String.wrap("b"));
        toAdd.set(2, String.wrap("c"));
        toAdd.set(3, String.wrap("d"));
        toAdd.set(4, String.wrap("e"));
        l.addArray(toAdd);
        this.assertTrue(l.size() == 5, String.wrap("Size should be 5"));
    }

    public removeIndex() : void {
        let val : String;
        let l : List<String> = new List<String>();
        l.add(String.wrap("a"));
        l.add(String.wrap("b"));
        l.add(String.wrap("c"));
        l.add(String.wrap("d"));
        l.add(String.wrap("e"));
        l.removeIndex(1);
        l.removeIndex(1);
        this.assertTrue(l.size() == 3, String.wrap("Size should be 3"));
        val = l.get(0);
        this.assertTrue(val.compareTo(String.wrap("a")) == 0, String.wrap("First element should be a"));
        val = l.get(1);
        this.assertTrue(val.compareTo(String.wrap("d")) == 0, String.wrap("Second element should be d"));
        val = l.get(2);
        this.assertTrue(val.compareTo(String.wrap("e")) == 0, String.wrap("Third element should be e"));
    }

    public removeRange() : void {
        let val : String;
        let l : List<String> = new List<String>();
        l.add(String.wrap("a"));
        l.add(String.wrap("b"));
        l.add(String.wrap("c"));
        l.add(String.wrap("d"));
        l.add(String.wrap("e"));
        l.removeRange(3, 2);
        this.assertTrue(l.size() == 3, String.wrap("Size should be 3"));
        val = l.get(0);
        this.assertTrue(val.compareTo(String.wrap("a")) == 0, String.wrap("First element should be a"));
        val = l.get(1);
        this.assertTrue(val.compareTo(String.wrap("b")) == 0, String.wrap("Second element should be b"));
        val = l.get(2);
        this.assertTrue(val.compareTo(String.wrap("c")) == 0, String.wrap("Third element should be c"));
    }
}


test('add', () => {
    new TestList().add();
});

test('addRange', () => {
    new TestList().addRange();
});

test('removeIndex', () => {
    new TestList().removeIndex();
});

test('removeRange', () => {
    new TestList().removeRange();
});

File: go/datastructure/testlist/testlist_test.go
package testlist

import (
    "testing"
    array "github.com/otterop/otterop/go/lang/array"
    generic "github.com/otterop/otterop/go/lang/generic"
    string "github.com/otterop/otterop/go/lang/string"
    testbase "github.com/otterop/otterop/go/test/testbase"
    "github.com/otterop/otterop/go/datastructure/list"
)


func NewTestList() *TestList {
    this := new(TestList)
    this.TestBase = testbase.NewTestBase()
    return this
}

type TestList struct {
    *testbase.TestBase
}




func (this *TestList) Add()  {
    var l *list.List[*string.String] = list.NewList[*string.String]()
    l.Add(string.Wrap(string.Literal("a")))
    l.Add(string.Wrap(string.Literal("b")))
    l.Add(string.Wrap(string.Literal("c")))
    l.Add(string.Wrap(string.Literal("d")))
    l.Add(string.Wrap(string.Literal("e")))
    this.AssertTrue(l.Size() == 5, string.Wrap(string.Literal("Size should be 5")))
}

func (this *TestList) AddRange()  {
    var genericString *generic.Generic[*string.String] = generic.NewGeneric[*string.String]()
    var genericT *string.String = genericString.Zero()
    var l *list.List[*string.String] = list.NewList[*string.String]()
    var toAdd *array.Array[*string.String] = array.NewArray(5, genericT)
    toAdd.Set(0, string.Wrap(string.Literal("a")))
    toAdd.Set(1, string.Wrap(string.Literal("b")))
    toAdd.Set(2, string.Wrap(string.Literal("c")))
    toAdd.Set(3, string.Wrap(string.Literal("d")))
    toAdd.Set(4, string.Wrap(string.Literal("e")))
    l.AddArray(toAdd)
    this.AssertTrue(l.Size() == 5, string.Wrap(string.Literal("Size should be 5")))
}

func (this *TestList) RemoveIndex()  {
    var val *string.String
    var l *list.List[*string.String] = list.NewList[*string.String]()
    l.Add(string.Wrap(string.Literal("a")))
    l.Add(string.Wrap(string.Literal("b")))
    l.Add(string.Wrap(string.Literal("c")))
    l.Add(string.Wrap(string.Literal("d")))
    l.Add(string.Wrap(string.Literal("e")))
    l.RemoveIndex(1)
    l.RemoveIndex(1)
    this.AssertTrue(l.Size() == 3, string.Wrap(string.Literal("Size should be 3")))
    val = l.Get(0)
    this.AssertTrue(val.CompareTo(string.Wrap(string.Literal("a"))) == 0, string.Wrap(string.Literal("First element should be a")))
    val = l.Get(1)
    this.AssertTrue(val.CompareTo(string.Wrap(string.Literal("d"))) == 0, string.Wrap(string.Literal("Second element should be d")))
    val = l.Get(2)
    this.AssertTrue(val.CompareTo(string.Wrap(string.Literal("e"))) == 0, string.Wrap(string.Literal("Third element should be e")))
}

func (this *TestList) RemoveRange()  {
    var val *string.String
    var l *list.List[*string.String] = list.NewList[*string.String]()
    l.Add(string.Wrap(string.Literal("a")))
    l.Add(string.Wrap(string.Literal("b")))
    l.Add(string.Wrap(string.Literal("c")))
    l.Add(string.Wrap(string.Literal("d")))
    l.Add(string.Wrap(string.Literal("e")))
    l.RemoveRange(3, 2)
    this.AssertTrue(l.Size() == 3, string.Wrap(string.Literal("Size should be 3")))
    val = l.Get(0)
    this.AssertTrue(val.CompareTo(string.Wrap(string.Literal("a"))) == 0, string.Wrap(string.Literal("First element should be a")))
    val = l.Get(1)
    this.AssertTrue(val.CompareTo(string.Wrap(string.Literal("b"))) == 0, string.Wrap(string.Literal("Second element should be b")))
    val = l.Get(2)
    this.AssertTrue(val.CompareTo(string.Wrap(string.Literal("c"))) == 0, string.Wrap(string.Literal("Third element should be c")))
}

func Test(t *testing.T) {
    t.Run("Add", func(t *testing.T) {
        test := NewTestList()
        test.SetGoTestingT(t)
        test.Add()
    })
    t.Run("AddRange", func(t *testing.T) {
        test := NewTestList()
        test.SetGoTestingT(t)
        test.AddRange()
    })
    t.Run("RemoveIndex", func(t *testing.T) {
        test := NewTestList()
        test.SetGoTestingT(t)
        test.RemoveIndex()
    })
    t.Run("RemoveRange", func(t *testing.T) {
        test := NewTestList()
        test.SetGoTestingT(t)
        test.RemoveRange()
    })
}
